src/controller/java/OTAProviderDelegateBridge.cpp

```diff
@@ -0,0 +1,375 @@
+        sendOTAQueryFailure(static_cast<uint8_t>(response.status));
+        return;
+    }
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+
+    jmethodID handleQueryImageMethod;
+    err = JniReferences::GetInstance().FindMethod(env, mOtaProviderDelegate, "handleQueryImage",
+                                                  "(IIJLjava/lang/Integer;Ljava/lang/String;Ljava/lang/Boolean;[B)Lchip/"
+                                                  "devicecontroller/OTAProviderDelegate$QueryImageResponse;",
+                                                  &handleQueryImageMethod);
+    VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(Controller, "Could not find handleQueryImage method"));
+
+    jobject boxedHardwareVersion = nullptr;
+    if (hardwareVersion.HasValue())
+    {
+        chip::JniReferences::GetInstance().CreateBoxedObject<jint>(
+            "java/lang/Integer", "(I)V", static_cast<jint>(hardwareVersion.Value()), boxedHardwareVersion);
+    }
+    jobject boxedLocation = nullptr;
+    if (location.HasValue())
+    {
+        boxedLocation = env->NewStringUTF(std::string(location.Value().data(), location.Value().size()).c_str());
+    }
+    jobject boxedRequestorCanConsent = nullptr;
+    if (requestorCanConsent.HasValue())
+    {
+        chip::JniReferences::GetInstance().CreateBoxedObject<jboolean>(
+            "java/lang/Boolean", "(Z)V", requestorCanConsent.Value() ? JNI_TRUE : JNI_FALSE, boxedRequestorCanConsent);
+    }
+    jobject boxedMetadataForProvider = nullptr;
+    if (metadataForProvider.HasValue())
+    {
+        jbyteArray boxedMetadataForProviderByteArray = env->NewByteArray(static_cast<jsize>(metadataForProvider.Value().size()));
+        env->SetByteArrayRegion(boxedMetadataForProviderByteArray, 0, static_cast<jsize>(metadataForProvider.Value().size()),
+                                reinterpret_cast<const jbyte *>(metadataForProvider.Value().data()));
+        boxedMetadataForProvider = boxedMetadataForProviderByteArray;
+    }
+
+    jobject jResponse =
+        env->CallObjectMethod(mOtaProviderDelegate, handleQueryImageMethod, static_cast<jint>(vendorId),
+                              static_cast<jint>(productId), static_cast<jlong>(softwareVersion), boxedHardwareVersion,
+                              boxedLocation, boxedRequestorCanConsent, boxedMetadataForProvider);
+    if (env->ExceptionCheck())
+    {
+        ChipLogError(Support, "Exception in call java method");
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        return;
+    }
+
+    bool hasUpdate = (jResponse != nullptr);
+
+    Commands::QueryImageResponse::Type response;
+    // If update is not available, return the delegate response
+    if (!hasUpdate)
+    {
+        response.status = OTAQueryStatus::kNotAvailable;
+        commandObj->AddResponse(cachedCommandPath, response);
+        sendOTAQueryFailure(static_cast<uint8_t>(response.status));
```

Comment:
For all error cases in these functions, should we call sendOTAQueryFailure across? In other places, it seems we siliently return or just show the error log?

Reply:
I called sendOTAQueryFailure only when the app called an intentional Error. (e.g., when there is no transmitted data, BDX Protocol is not supported).
Is it correct to call it in other cases as well?

Reply:
The other side would timeout if there is no response here, which means client stucks somehow...
Also it would be good to report the error to controller application as well, instead of silently skipping it.

I would prefer generating right error to notify if we can.

Reply:
modified. All cases will be called callback function,.